### DISCUSSION TITLE
Удалить из фонотеки прослушанные альбомы и плейлисты

### DISCUSSION THREAD

**Discussion by Bedmit** [Chimildic/goofy#127] [Помощь с алгоритмом] [answered]
Возможна ли проверка всех избранных альбомов фонотеки на предмет того, что треки в них уже в истории прослушивания и/или в любимых треках, и удаления их из фонотеки в этом случае?

Получить треки всех избранных альбомов и отнять от них  историю и лайки, это понятно. Но как определить какие альбомы остались без треков, чтобы удалить их из фонотеки?

**Comment by Chimildic** [comment, marked as answer]
Обновил библиотеку. Сделал функцию получения сохраненных альбомов публичной [getSavedAlbums](https://chimildic.github.io/goofy/#/func?id=getsavedalbums).

Переменная `filteredAlbums` будет содержать альбомы, в которых не осталось треков после фильтра по `banTracks`. Отдав результат в `Library.deleteAlbums`, эти альбомы удалятся. Закомментировал строчку поскольку деструктивное действие.
```js
let banTracks = Combiner.push(Source.getSavedTracks(), RecentTracks.get());
let filteredAlbums = Source.getSavedAlbums().filter(album => {
  Filter.removeTracks(album.tracks.items, banTracks);
  if (album.tracks.items.length == 0) {
    console.log('Альбом', album.name, 'от', album.artists[0].name, 'полностью прослушан');
    return true;
  }
});
// Library.deleteAlbums(filteredAlbums);
```

**Reply by Bedmit** [reply]
Как всегда спасибо. То что надо.

**Comment by Bedmit** [comment]
А аналогично можно с избранными плейлистами делать?
Даже не со всеми, чтобы сильно не загружать, а теми которые не обновляются какой-то период. Т.е. получаем плейлисты не обновлявшиеся последние пол года, проверяем все ли треки в них прослушаны/лайкнуты. Если все - удаляем плейлист из фонотеки.

_Уточню. Это все совершенно не обязательно, если требует каких-то серьезных телодвижений. А то мне уже неудобно, что ни попрошу - тут же под мои хотелки вся библиотека перекраивается ))_

**Comment by Chimildic** [comment]
Вновь обновил библиотеку для работы с избранными плейлистами. Описанный алгоритм возможен, но есть нюансы реализации:

- Сам плейлист не имеет даты обновления, поэтому нужно собирать его треки. В целом не очень накладно даже при большом количестве подписок. 
- Треки не каждого плейлиста имеют дату добавления. Например, в личных миксах или радио ее нет.  
- Избранные _топы за год_ очень вероятно будут полностью прослушаны. 
- Полгода большой срок, некоторые прослушивания могут выпасть из лимита истории (20 тысяч).

Вижу как минимум два решения:
1. Вместо отписки делать сводку. К примеру плейлист из непрослушанного или письмо на почту (Apps Script умеет). И уже вручную решать что удалять. 
2. Указать белый список, который не будет проверятся (`exclude`)

Запустив следующий код без изменений, после обновления, в консоле будут имена плейлистов, которые _попали под раздачу_. Поможет определиться с белым списком. Обновил код под алгоритм из комментариев.
```js
  let banTracks = Combiner.push(Source.getSavedTracks(), RecentTracks.get());
  let playlists = Source.getFollowedTracks({
    isFlat: false,
    exclude: [
      // { name: '', id: '' },
    ]
  });
  let logs = [];
  let filteredPlaylists = playlists.filter(p => {
    let copyTracks = Selector.sliceCopy(p.tracks.items);
    Filter.rangeDateRel(copyTracks, 182);
    if (copyTracks.length > 0) {
      return false;
    }
    logs.push(`\nПлейлист не обновлялся минимум полгода ${p.name} (${p.id})`);
    Filter.removeTracks(p.tracks.items, banTracks);
    if (p.tracks.items.length == 0) {
      logs.push('Все треки прослушаны')
      return true;
    }
    logs.push(`${p.tracks.items.length} - треков нет в истории`);
  });
  console.log(logs.join('\n'))
  // Library.unfollowPlaylists(filteredPlaylists);
```

> Нельзя удалить личный плейлист. Только отписаться от чужих (снять лайк). Возможен костыль: генерировать себе плейлисты под другим аккаунтом.

**Reply by Bedmit** [reply]
Добрался до тестирования. То ли я что-то не так понимаю, то ли что-то не так как надо работает.
В результат выдало четыре плейлиста, которые действительно уже год не обновлялись. Но  они еще не прослушаны. 
На всякий случай взял один из них, отнял историю и лайки, и получил 30 треков. Такое впечатление, что если первое условие - плейлист больше полу года не обновляется, верно, то второе - все треки плейлиста должны быть прослушаны/пролайканы, уже не проверяется.

**Reply by Chimildic** [reply]
Я об этом и написал
> Сочетание фильтров также под вопросом. Сейчас это условие, что треки за последние полгода полностью прослушаны. Более старые треки вообще не участвуют в проверке, среди них могут быть непрослушанные.

**Reply by Bedmit** [reply]
А можно немного не так. Проверить плейлисты на предмет того, что они не обновлялись пол года, а из них выявить те которые полностью прослушаны? Роли не играет за какое время они прослушаны, главное что их треки есть в истории или в лайках.

Суть та же, что и с альбомами. Я добавляю в избранное плейлисты/альбомы, что бы их треки составляли плейлист дня. Прослушивая эти треки, я как бы "выжимаю" эти плейлисты/альбомы. То что понравилось - добавлено в лайки, то что не понравилось и не добавлено в лайки - мне больше не нужно. "Выжатый" плейлист/альбом, в котором не осталось прослушанных треков - удаляется из избранного. Исключение составляют периодически обновляемые плейлисты. Они остаются в ожидании появления новых треков. Но есть статичные плейлисты, которые создаются один раз по случаю и больше не обновляются. Вот эти плейлисты нужно удалять. Примерный срок "необновления" - пол года. Это ориентировочный срок, просто показывающий, что плейлист в принципе не обновляется и наиболее вероятно не будет.

**Reply by Chimildic** [reply]
> создаются один раз по случаю и больше не обновляются. Вот эти плейлисты нужно удалять.

Плейлисты нельзя удалять. Только отписаться (снять лайк). Получается автор плейлиста не владелец аккаунта? Хм, тогда их можно генерировать под другим аккаунтом) 

Если все ещё подходит, то я не до конца понимаю связь прослушиваний и даты в полгода. Можно сформулировать предложениями "если * и * то *". Почти готовый код получится)

**Reply by Bedmit** [reply]
> Плейлисты нельзя удалять

Удалять из фонотеки. Как бы название темы )



> Можно сформулировать предложениями "если * и * то *

Проверить избранные плейлисты.
Если среди них есть плейлисты не обновляемые больше полу года, то проверить есть ли в этих плейлистах треки не из истории прослушивания и любимых треков.
Если нет, то удалить эти плейлисты из фонотеки.

Сейчас, если удалить из вышеуказанного кода строчку `Filter.rangeDateRel(p.tracks.items, 182);` получится примерно нужный результат - полностью прослушанные плейлисты. Но часть плейлистов вероятнее всего еще может обновиться, т.к. они периодически обновляются. Например раз в три-четыре месяца. Такие плейлисты удалять из фонотеки не нужно. Вот их я и хочу выявить. Вернее  выявить и удалить те, которые вряд ли обновятся, т.к. давно не обновляются (для себя определил срок в пол года).

**Reply by Chimildic** [reply]
> Удалять из фонотеки. Как бы название темы

Просто сформулировано про "создаются". Решил уточнить, что личные нельзя удалить. Только отписываться от избранного.

Новое описание вроде прояснило алгоритм. Чуть позже напишу его.

**Reply by Chimildic** [reply]
Обновил основной пост. Проблема радио без дат и лимита истории остается актуальной.

**Reply by Bedmit** [reply]
Спасибо. Я подозревал, что нужно нужно было разделить этот цикл на два. Но сомневался )
Только одно уточнение. Почему закомментированная строка удаляющая плейлист, находится под этими циклами? Не будет ли она его удалять и в случае наличия не прослушанных треков, если ее раскомментировать? Может ее сюда? -
```js
console.log(`Плейлист не обновлялся минимум полгода ${p.name} (${p.id})`);
    Filter.removeTracks(p.tracks.items, banTracks);
    if (p.tracks.items.length == 0) {
      console.log('Все треки прослушаны')

      Library.unfollowPlaylists(filteredPlaylists);
      return true;
    }
    console.log(`${p.tracks.items.length} - треков нет в истории`);
```
На счет лимитов понял. Возможно у кого-то это проявится, но в  моем частном случае использования, не смущает. Миксы и радио в избранное не добавляю. Пол года у меня вроде как помещается в 20000 треков. Видимо из-за удаления в ней дублей. Отчасти поэтому и выбран такой размер условия.

**Reply by Chimildic** [reply]
В целом это больше стиль, чем необходимость. Меньше вложенность ветвлений, больше вариантов к расширению.
По плейлистам проходит `filter` чья задача вернуть в переменную `filteredPlaylists` только то, что вернуло `true` внутри него. Это произойдет только при попадании в эту ветку. Во всех остальных случаях плейлист отбрасывается. 
```js
if (p.tracks.items.length == 0) {
    console.log('Все треки прослушаны')
    return true;
}
```
То есть количество таких сообщений в консоли эквивалент найденных плейлистов. Можно убедиться, поставив точку остановки на последнюю строчку функции `}` и запустить отладку.

**Reply by Bedmit** [reply]
Примерно понял )) Спасибо.
Добавил еще удаление недоступных треков 
```js
Filter.removeTracks(p.tracks.items, banTracks);
Filter.removeUnavailable(p.tracks.items);
```
Только теперь из-за них не очень удобно просматривать сообщения в консоли. Я так понимаю сообщения о недоступных треках в консоли по умолчанию. А можно это отключить по желанию? Мне вообще в данном случае только сообщения об удаляемых плейлистах полезны.

**Reply by Chimildic** [reply]
Как вариант, собирать лог в массив и вывести в конце. Обновил пост. 
Если неудобно, придется к библиотеке добавлять аргумент выбора.

**Reply by Bedmit** [reply]
Да не критично, можно и так.
Разве что когда совсем идеи иссякнут и нечего будет обновить в библиотеке ))
